rfc/custom-modules-v1.md

```diff
@@ -0,0 +1,568 @@
+- The current module system does not provide an intuitive way to create or modify OpenTelemetry data, logs for different parts of the gateway lifecycle.
+- The current module system does not provide a way to interact with the parsed, normalized, and planned GraphQL operation in order to implement custom logic.
+- The current module system does not provide a way to hook into authorization and authentication logic in the gateway.
+- The current module system does not provide a way to hook into the usage of GraphQL directives in the operation definition or subgraph schema.
+
+Ultimately, custom modules must be self-contained, composable and testable. They should provide a clear API for developers to interact with the gateway and subgraph lifecycle and implement custom logic without having to understand the internal workings of the router or advanced Go programming concepts.
+To briefly explain the decision to use Go as the language for the module system, we have chosen Go because it is a simple and easy-to-learn language that is widely used in the infrastructure and cloud-native ecosystem. You can build custom integration on top production-grade SDK of AWS, GCP and the community without re-implementing them from scratch. It superiors to scripting languages like Rhai or cross-compiling WebAssembly because it can be easily debugged, profiled, and tested with any modern IDE (VsCode, Goland, etc.). Not part of this RFC, are our ambitions to make the workflow as smooth as possible with a CLI tool that can scaffold, test, and deploy custom modules. In the future, custom modules could be published to a central registry and shared with the community. A brief overview of the workflow is provided at the end of this RFC.
+
+As powerful as the new module becomes, it is important to move basic and common functionality into the core of the router because building and maintaining custom modules should be a last resort. The router should provide a rich set of features out of the box that cover the most common use cases. Custom modules should be reserved for advanced or highly specific use cases that cannot be achieved with the built-in features of the router. Integration with third-party services, custom authentication, and advanced telemetry are examples of use cases that are well-suited for custom modules.
+
+## Proposal
+
+A developer can implement a custom module by creating a struct that implements one or more of the following interfaces:
+
+- `GatewayHooks`: Provides hooks for the gateway lifecycle, including request and response handling.
+- `SubgraphHooks`: Provides hooks for subgraph requests and responses.
+- `ApplicationHooks`: Provides hooks for the application lifecycle, including startup, shutdown, and error handling.
+- `AuthenticationHooks`: Provides hooks for authentication and authorization logic.
```

Comment:
Should we distinguish Authentication & Authorization into different hooks?
AuthN typically comes at the Gateway level, AuthZ at the Subgraph or Field level.

Reply:
Yes. I separated them.